test/controllers/agent/TaxYearsControllerSpec.scala

```diff
@@ -0,0 +1,175 @@
+
+class TaxYearsControllerSpec extends TestSupport
+  with MockIncomeSourceDetailsService
+  with MockFrontendAuthorisedFunctions
+  with MockItvcErrorHandler
+  with MockCalculationService
+  with MockTaxYears
+  with FeatureSwitching {
+
+  override def beforeEach(): Unit = {
+    super.beforeEach()
+    disable(AgentViewer)
+  }
+
+  trait Setup {
+    val controller = new TaxYearsController(taxYears, mockAuthService, mockCalculationService, mockIncomeSourceDetailsService)(
+      app.injector.instanceOf[FrontendAppConfig],
+      app.injector.instanceOf[MessagesControllerComponents],
+      app.injector.instanceOf[ExecutionContext],
+      mockItvcErrorHandler
+    )
+  }
+
+  "show" when {
+    "the user is not authenticated" should {
+      "redirect them to sign in" in new Setup {
+        setupMockAgentAuthorisationException()
+
+        val result: Future[Result] = controller.show()(fakeRequestWithActiveSession)
+
+        status(result) shouldBe SEE_OTHER
+        redirectLocation(result) shouldBe Some(controllers.routes.SignInController.signIn().url)
+      }
+    }
+    "the user has timed out" should {
+      "redirect to the session timeout page" in new Setup {
+        setupMockAgentAuthorisationException(exception = BearerTokenExpired())
+
+        val result: Future[Result] = controller.show()(fakeRequestWithTimeoutSession)
+
+        status(result) shouldBe SEE_OTHER
+        redirectLocation(result) shouldBe Some(controllers.timeout.routes.SessionTimeoutController.timeout().url)
+      }
+    }
+    "the user does not have an agent reference number" should {
+      "return Ok with technical difficulties" in new Setup {
+        setupMockAgentAuthRetrievalSuccess(testAgentAuthRetrievalSuccessNoEnrolment)
+        mockShowOkTechnicalDifficulties()
+
+        val result: Future[Result] = controller.show()(fakeRequestWithActiveSession)
+
+        status(result) shouldBe OK
+        contentType(result) shouldBe Some(HTML)
+      }
+    }
+    "the agent viewer feature switch is disabled" should {
+      "return Not Found" in new Setup {
+        setupMockAgentAuthRetrievalSuccess(testAgentAuthRetrievalSuccess)
+        mockNotFound()
+
+        val result: Future[Result] = controller.show()(fakeRequestConfirmedClient())
+
+        status(result) shouldBe NOT_FOUND
+        contentType(result) shouldBe Some(HTML)
+      }
+    }
+    "the agent viewer feature switch is enabled" when {
+      "there was a problem retrieving income source details for the user" should {
+        "throw an internal server exception" in new Setup {
+          enable(AgentViewer)
+          setupMockAgentAuthRetrievalSuccess(testAgentAuthRetrievalSuccess)
+          mockErrorIncomeSource()
+          mockShowInternalServerError()
+
+          intercept[InternalServerException](await(controller.show()(fakeRequestConfirmedClient())))
+            .message shouldBe "[HomeController][getMtdItUserWithIncomeSources] IncomeSourceDetailsModel not created"
```

Comment:
Suspicious discrepancy. Should the message start with `[TaxYearsController][show]` rather than `[HomeController][getMtdItUserWithIncomeSources]`?